driver.go

```diff
@@ -33,67 +36,151 @@ import (
 // Use fully qualified string:
 //
 // Example: projects/$PROJECT/instances/$INSTANCE/databases/$DATABASE
 func (d *Driver) Open(name string) (driver.Conn, error) {
-	return openDriverConn(context.Background(), d, name)
+	c, err := newConnector(d, name)
+	if err != nil {
+		return nil, err
+	}
+	return openDriverConn(context.Background(), c)
 }
 
 func (d *Driver) OpenConnector(name string) (driver.Connector, error) {
-	return &connector{
-		driver: d,
-		name:   name,
+	return newConnector(d, name)
+}
+
+type connectorConfig struct {
+	host     string
+	project  string
+	instance string
+	database string
+	params   map[string]string
+}
+
+func extractConnectorConfig(dsn string) (connectorConfig, error) {
+	match := dsnRegExp.FindStringSubmatch(dsn)
+	matches := make(map[string]string)
+	for i, name := range dsnRegExp.SubexpNames() {
+		if i != 0 && name != "" {
+			matches[name] = match[i]
+		}
+	}
+	paramsString := matches["PARAMSGROUP"]
+	params, err := extractConnectorParams(paramsString)
+	if err != nil {
+		return connectorConfig{}, err
+	}
+
+	return connectorConfig{
+		host:     matches["HOSTGROUP"],
+		project:  matches["PROJECTGROUP"],
+		instance: matches["INSTANCEGROUP"],
+		database: matches["DATABASEGROUP"],
+		params:   params,
 	}, nil
 }
 
+func extractConnectorParams(paramsString string) (map[string]string, error) {
+	params := make(map[string]string)
+	if paramsString == "" {
+		return params, nil
+	}
+	keyValuePairs := strings.Split(paramsString, ";")
+	for _, keyValueString := range keyValuePairs {
+		keyValue := strings.SplitN(keyValueString, "=", 2)
+		if keyValue == nil || len(keyValue) != 2 {
+			return nil, fmt.Errorf("invalid connection property: %s", keyValueString)
+		}
+		params[keyValue[0]] = keyValue[1]
+	}
+	return params, nil
+}
+
 type connector struct {
-	driver *Driver
-	name   string
+	driver          *Driver
+	connectorConfig connectorConfig
+
+	// config represents the optional advanced configuration to be used
+	// by the Google Cloud Spanner client.
+	config spanner.ClientConfig
+
+	// options represent the optional Google Cloud client options
+	// to be passed to the underlying client.
+	options []option.ClientOption
+}
+
+func newConnector(d *Driver, dsn string) (*connector, error) {
+	connectorConfig, err := extractConnectorConfig(dsn)
+	if err != nil {
+		return nil, err
+	}
+	opts := make([]option.ClientOption, 0)
+	if connectorConfig.host != "" {
+		opts = append(opts, option.WithEndpoint(connectorConfig.host))
+	}
+	if strval, ok := connectorConfig.params["useplaintext"]; ok {
+		if val, err := strconv.ParseBool(strval); err == nil && val {
+			opts = append(opts, option.WithGRPCDialOption(grpc.WithInsecure()), option.WithoutAuthentication())
+		}
+	}
+	config := spanner.ClientConfig{
+		SessionPoolConfig: spanner.DefaultSessionPoolConfig,
+	}
+	return &connector{
+		driver:          d,
+		connectorConfig: connectorConfig,
+		config:          config,
```

Comment:
can we have a better name for `config`? We have `connectorConfig` and `config` in this struct. It is unclear what `config` is for.

Reply:
Good point, changed to `spannerClientConfig`.